fix(router): update type for routerLink to include null and undefined

atscott opened the pull request:
PR #13380 added support for `null` and `undefined` but the type on the parameter was not updated. This would result in a compilation error if `fullTemplateTypeCheck` is enabled.

Fixes #36544

atscott committed:
fix(router): update type for routerLink to include null and undefined

PR #13380 added support for `null` and `undefined` but the type on the parameter was not updated.
This would result in a compilation error if `fullTemplateTypeCheck` is enabled.
Fixes #36544

AndrewKushnir reviewed (approved):
LGTM 👍

atscott commented:
[presubmit](https://test.corp.google.com/ui#id=OCL:310590576:BASE:310590602:1588961276540:a823eae6)

IgorMinar commented on packages/router/src/directives/router_link.ts:
```diff
@@ -229,7 +229,7 @@ export class RouterLinkWithHref implements OnChanges, OnDestroy {
 
   @Input()
-  set routerLink(commands: any[]|string) {
+  set routerLink(commands: any[]|string|null|undefined) {
```
can you please add docs and document what will be the resulting behavior of passing `null` and `undefined` into `[router-link]`? thanks!

IgorMinar reviewed (changes requested):
can you do a light past to add the docs? Please don’t spend too much time on it, even a single sentence per property is better than nothing.

https://next.angular.io/api/router/RouterLink
https://next.angular.io/api/router/RouterLinkWithHref

Thanks!

atscott committed:
docs(router): add docs for RouterLink inputs

The RouterLink and RouterLinkWithHref inputs do not have any docs. This comment adds jsdoc comments to the inputs.

IgorMinar commented on packages/router/src/directives/router_link.ts:
```diff
@@ -115,19 +115,49 @@ import {UrlTree} from '../url_tree';
   // TODO(issue/24571): remove '!'.
   @Input() queryParams!: {[k: string]: any};
+  /**
+   * Passed to `Router.createUrlTree` as part of the `NavigationExtras`.
```
please use `#` instead of `.` in this case. The community convention is that`#` is used for instance methods and `.` is for static methods.
```suggestion
   * Passed to `Router#createUrlTree` as part of the `NavigationExtras`.
```

mary-poppins commented:
You can preview 62a6755 at https://pr37018-62a6755.ngbuilds.io/.
You can preview 10fc47f at https://pr37018-10fc47f.ngbuilds.io/.
You can preview 92beec0 at https://pr37018-92beec0.ngbuilds.io/.
You can preview 1719671 at https://pr37018-1719671.ngbuilds.io/.
You can preview 1e66166 at https://pr37018-1e66166.ngbuilds.io/.
You can preview 01c5d4f at https://pr37018-01c5d4f.ngbuilds.io/.

IgorMinar commented on packages/router/src/directives/router_link.ts:
```diff
@@ -115,19 +115,49 @@ import {UrlTree} from '../url_tree';
+  /**
+   * Passed to `Router.createUrlTree` as part of the `NavigationExtras`.
+   * @see [NavigationExtras](https://angular.io/api/router/NavigationExtras#navigationextras)
+   * @see [Router.createUrlTree](https://angular.io/api/router/Router#createurltree)
```
Please don't use absolute links anywhere in angular.io as this will break navigation flow when a user is visiting a versioned deployment of the docs (e.g. v9.angular.io). In fact, all that you should need here is `Router#createUrlTree`.

See an existing example: https://github.com/angular/angular/blob/22d80a6780c235b2e8c64b23ebe7dd49c49f9aae/packages/core/src/interface/lifecycle_hooks.ts#L13-L18

IgorMinar commented on packages/router/src/directives/router_link.ts:
```diff
@@ -115,19 +115,49 @@ import {UrlTree} from '../url_tree';
+   * Passed to `Router.createUrlTree` as part of the `NavigationExtras`.
+   * @see [NavigationExtras](https://angular.io/api/router/NavigationExtras#navigationextras)
+   * @see [Router.createUrlTree](https://angular.io/api/router/Router#createurltree)
+   */
```
btw I'm not sure if you know, but if you add `aio: preview` label to this PR you get a comment posted on the PR by @mary-poppins - the comment then contains a link to the preview of the docs at each SHA in this PR.

You can then navigate to that URL and see what your change will look like on angular.io  without you having to build it locally.

Example link for this particular change: https://pr37018-01c5d4f.ngbuilds.io/api/router/RouterLink#properties

atscott replied:
I learned this after see you add the label to the PR. Very useful, thanks for pointing that out.

IgorMinar commented on packages/router/src/directives/router_link.ts:
```diff
@@ -228,15 +300,25 @@ export class RouterLinkWithHref implements OnChanges, OnDestroy {
   }
 
+  /**
+   * @deprecated 4.0.0 use `queryParamsHandling` instead.
```
```suggestion
   * @deprecated As of Angular v4.0 use `queryParamsHandling` instead.
```

IgorMinar reviewed (approved):
LGTM after the docs are updated. Thank you!

Reviewed-for: public-api

atscott committed:
fixup! docs(router): add docs for RouterLink inputs

mary-poppins commented:
You can preview 266b6e0 at https://pr37018-266b6e0.ngbuilds.io/.

mary-poppins commented:
You can preview 06e8856 at https://pr37018-06e8856.ngbuilds.io/.

mary-poppins commented:
You can preview 7a32731 at https://pr37018-7a32731.ngbuilds.io/.

mary-poppins commented:
You can preview 5f66203 at https://pr37018-5f66203.ngbuilds.io/.

mary-poppins commented:
You can preview f2ecb52 at https://pr37018-f2ecb52.ngbuilds.io/.

atscott commented:
merge assistance: I'm the owner of the router. LGTMs from Igor and Andrew Kushnir should cover required approvals.
